azkaban-db/src/main/java/azkaban/db/MySQLDataSource.java

```diff
@@ -58,18 +65,126 @@ public synchronized Connection getConnection() throws SQLException {
          * Every Attempt generates a thread-hanging-time, about 75 seconds, which is hard coded, and can not be changed.
          */
         connection = createDataSource().getConnection();
-        if (connection != null) {
+
+        /**
+         * If connection is null or connection is read only, retry to find available connection.
+         * When DB fails over from master to slave, master is set to read-only mode. We must keep
+         * finding correct data source and sql connection.
+         */
+        if (connection == null || isReadOnly(connection)) {
+          throw new SQLException("Failed to find DB connection Or connection is read only. ");
+        } else {
           return connection;
         }
       } catch (final SQLException ex) {
-        logger.error(
-            "Failed to find DB connection. waits 1 minutes and retry. No.Attempt = " + retryAttempt,
-            ex);
-      } finally {
+
+        logger.error( "Failed to find write-enabled DB connection. Wait 1 minutes and retry."
+            + " No.Attempt = " + retryAttempt, ex);
+        /**
+         * When database is completed down, DB connection fails to be fetched immediately. So we need
+         * to hang 1 minute for retry.
+         */
+        sleep(1000L * 60);
         retryAttempt++;
+        this.dataSource = null;
       }
     }
-    return null;
+    return connection;
+  }
+
+  /**
+   * This method keeps the original functionality from parent class. The difference is that we want to
```

Comment:
This doesn't seem to be ideal. Is there no better way?

Reply:
Have you tried calling invalidateConnection and maybe clear the connection pool too when the connection is readonly to force DBCP to reconnect to the new host?
https://commons.apache.org/proper/commons-dbcp/apidocs/src-html/org/apache/commons/dbcp2/BasicDataSource.html#line.1971

http://commons.apache.org/proper/commons-pool/api-2.4.2/org/apache/commons/pool2/impl/GenericObjectPool.html#clear--
https://commons.apache.org/proper/commons-dbcp/apidocs/org/apache/commons/dbcp2/BasicDataSource.html#getConnectionPool--

Reply:
Another possibility is to wrap the datasource and proxy calls to it.
https://stackoverflow.com/questions/5007371/refresh-datasource-using-springdbcp